### PR TITLE
Introduce 'var' declaration to Assignment statements.

### PR DESCRIPTION
**Pull request by djkevincr**

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/statements/AssignStmt.java [hunk 1]
```diff
@@ -53,4 +54,17 @@ public void setRExpr(Expression rhsExpr) {
     public void accept(NodeVisitor visitor) {
         visitor.visit(this);
     }
+
+    public void setRhsExpr(Expression rhsExpr) {
+        this.rhsExpr = rhsExpr;
+    }
+
+    public boolean isVarDeclaration() {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/statements/AssignStmt.java]: Shall we rename this to isDeclaredWithVar

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/CodeGenerator.java [hunk 2]
```diff
@@ -670,7 +670,15 @@ public void visit(ParameterDef parameterDef) {
 
     @Override
     public void visit(VariableDef variableDef) {
-
+        MemoryLocation memoryLocation = variableDef.getMemoryLocation();
+        if (memoryLocation instanceof StackVarLocation || memoryLocation instanceof WorkerVarLocation) {
+            int lvIndex = getNextIndex(variableDef.getType().getTag(), lvIndexes);
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/CodeGenerator.java]: This code block and the code block in 'visit(VariableDefStmt varDefStmt)' can be merged.

Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/CodeGenerator.java]: MemoryLocation memoryLocation = varDefStmt.getVariableDef().getMemoryLocation();
         if (memoryLocation instanceof StackVarLocation || memoryLocation instanceof WorkerVarLocation) {
             OpcodeAndIndex opcodeAndIndex = getOpcodeAndIndex(variableDef.getType().getTag(),
                     InstructionCodes.ISTORE, lvIndexes);
             opcode = opcodeAndIndex.opcode;
             lvIndex = opcodeAndIndex.index;
             stackVarLocation = new StackVarLocation(lvIndex);
             variableDef.setMemoryLocation(stackVarLocation);
             if (rhsExpr != null) {
                 emit(opcode, rhsExpr.getTempOffset(), lvIndex);
             }
 
             LocalVariableInfo localVarInfo = getLocalVarAttributeInfo(variableDef);
             currentlLocalVarAttribInfo.addLocalVarInfo(localVarInfo);
         } else {
             // TODO
         }

Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/CodeGenerator.java]: The above block without the instruction emit

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 3]
```diff
@@ -1105,6 +1105,26 @@ public void visit(ParameterDef paramDef) {
 
     @Override
     public void visit(VariableDef varDef) {
+        // Resolves the type of the variable
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: This code block should not be in this method. This logic is not generic to all variable definitions

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 4]
```diff
@@ -1173,18 +1193,46 @@ public void visit(VariableDefStmt varDefStmt) {
...
+                variableDef.accept(this);
+            }
+        }
         visitLExprsOfAssignment(assignStmt, lExprs);
 
         Expression rExpr = assignStmt.getRExpr();
         if (rExpr instanceof FunctionInvocationExpr || rExpr instanceof ActionInvocationExpr) {
             rExpr.accept(this);
+            if (assignStmt.isVarDeclaration()) {
+                if (rExpr instanceof FunctionInvocationExpr) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Why do you need this if condition? The 'getTypes' method comes from the "CallableUnitInvocationExpr" interface and both 'FunctionInvocationExpr' and 'ActionInvocationExpr' classes implements that interface.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 5]
```diff
@@ -1173,18 +1193,46 @@ public void visit(VariableDefStmt varDefStmt) {
     @Override
     public void visit(AssignStmt assignStmt) {
         Expression[] lExprs = assignStmt.getLExprs();
+        if (assignStmt.isVarDeclaration()) {
+            for (Expression expr : lExprs) {
+                if (!(expr instanceof VariableRefExpr)) {
+                    BLangExceptionHelper.throwSemanticError(assignStmt, SemanticErrors.INVALID_VAR_ASSIGNMENT);
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Use 'getSemanticException' method and then throw the exception. We need to refactor whole class later.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 6]
```diff
@@ -1173,18 +1193,46 @@ public void visit(VariableDefStmt varDefStmt) {
...
+                }
+                VariableRefExpr refExpr = (VariableRefExpr) expr;
+                Identifier identifier = new Identifier(refExpr.getVarName());
+                SymbolName symbolName = new SymbolName(identifier.getName());
+                VariableDef variableDef = new VariableDef(refExpr.getNodeLocation(),
+                        refExpr.getWhiteSpaceDescriptor(), identifier,
+                        new SimpleTypeName("var"), symbolName, currentScope);
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: This is wrong IMO. "var" is not a type. Please suggest an alternative solution

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 7]
```diff
@@ -1173,18 +1193,46 @@ public void visit(VariableDefStmt varDefStmt) {
...
+            }
             checkForMultiAssignmentErrors(assignStmt, lExprs, (CallableUnitInvocationExpr) rExpr);
             return;
         }
 
         if (lExprs.length > 1 && (rExpr instanceof TypeCastExpression || rExpr instanceof TypeConversionExpr)) {
             ((AbstractExpression) rExpr).setMultiReturnAvailable(true);
             rExpr.accept(this);
+            if (assignStmt.isVarDeclaration()) {
+                if (rExpr instanceof TypeCastExpression) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Why do you need this if condition? The 'getTypes' method comes from the "ExecutableMultiReturnExpr" interface and both 'FunctionInvocationExpr' and 'ActionInvocationExpr' classes implements that interface.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 8]
```diff
@@ -1202,6 +1253,10 @@ public void visit(AssignStmt assignStmt) {
 
         visitSingleValueExpr(rExpr);
         BType rhsType = rExpr.getType();
+        if (assignStmt.isVarDeclaration()) {
+            assignVariableRefTypes(new Expression[]{lExpr}, new BType[]{rhsType});
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Do you really need to invoke 'assignVariableRefTypes' method here? :) 

I would just set the type then and there

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 9]
```diff
@@ -1213,6 +1268,13 @@ public void visit(AssignStmt assignStmt) {
         }
     }
 
+    private static void assignVariableRefTypes(Expression[] expr, BType[] returnTypes) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Please move this method to the private method section of this class.

Comment (djkevincr) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: +1

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java (hunk 8 above)
Comment (djkevincr) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: +1

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java (hunk 7 above)
Comment (djkevincr) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: +1

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java (hunk 4 above)
Comment (djkevincr) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: +1

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/statements/AssignStmt.java (hunk 1 above)
Comment (djkevincr) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/statements/AssignStmt.java]: +1

**Pull request merged**